File: sdks/php/test/Model/SubFormFieldsPerDocumentTest.php
<?php

declare(strict_types=1);

namespace Dropbox\Sign\Test\Model;

use Dropbox\Sign\Model\SignatureRequestSendRequest;
use Dropbox\Sign\Test\SignTestCase;
use Dropbox\Sign\Test\TestUtils;

class SubFormFieldsPerDocumentTest extends SignTestCase
{
    /**
     * @dataProvider providerSubFormFieldsPerDocumentBase
     */
    public function testSubFormFieldsPerDocumentBase(
        string $type,
        array $form_field
    ) {
        $data = [
            'form_fields_per_document' => [$form_field],
        ];

        $obj = SignatureRequestSendRequest::init($data);

        $field = $obj->getFormFieldsPerDocument()[0];

        $this->assertInstanceOf("\\Dropbox\\Sign\\Model\\{$type}", $field);
        $this->assertEquals(
            $data['form_fields_per_document'],
            json_decode(json_encode($obj->getFormFieldsPerDocument()), true)
        );
    }

    /**
     * @dataProvider providerSubFormFieldsPerDocumentBase
     */
    public function testEmptyArrayReturnsNullValue(
        string $type,
        array $form_field
    ) {
        $data = [
            'form_fields_per_document' => [
                $form_field,
                [],
            ],
        ];

        $expected = [
            $form_field,
            null,
        ];

        $obj = SignatureRequestSendRequest::init($data);

        $field = $obj->getFormFieldsPerDocument()[0];

        $this->assertInstanceOf("\\Dropbox\\Sign\\Model\\{$type}", $field);
        $this->assertEquals(
            $expected,
            json_decode(json_encode($obj->getFormFieldsPerDocument()), true)
        );
    }

    /**
     * @dataProvider providerSubFormFieldsPerDocumentBase
     */
    public function testSignerAllowsInt(
        string $type,
        array $form_field
    ) {
        $form_field['signer'] = 1234;
        $data = ['form_fields_per_document' => [$form_field, []]];
        $expected_signer = '1234';

        $obj = SignatureRequestSendRequest::init($data);
        $result = $obj->getFormFieldsPerDocument()[0];
        $this->assertEquals($expected_signer, $result->getSigner());
    }

    /**
     * @dataProvider providerSubFormFieldsPerDocumentBase
     */
    public function testSignerAllowsString(
        string $type,
        array $form_field
    ) {
        $form_field['signer'] = 'sender';
        $data = ['form_fields_per_document' => [$form_field, []]];
        $expected_signer = 'sender';

        $obj = SignatureRequestSendRequest::init($data);
        $result = $obj->getFormFieldsPerDocument()[0];
        $this->assertEquals($expected_signer, $result->getSigner());
    }

    public function providerSubFormFieldsPerDocumentBase(): iterable
    {
        $fixtures = TestUtils::getFixtureData('SubFormFieldsPerDocument');
        foreach ($fixtures as $type => $data) {
            yield [
                'type' => $type,
                'data' => $data,
            ];
        }
    }
}

File: sdks/java-v1/src/test/java/com/dropbox/sign/SubFormFieldsPerDocumentTest.java
package com.dropbox.sign;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.dropbox.sign.model.SubFormFieldsPerDocumentBase;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Iterator;
import java.util.Map;
import org.junit.jupiter.api.Test;

public class SubFormFieldsPerDocumentTest {
    public final String packageNamePrefix = "com.dropbox.sign.model.";

    @Test
    public void testSubFormFieldsPerDocumentBase() throws Exception {
        ObjectMapper mapper = JSON.getDefault().getMapper();
        JsonNode content =
                mapper.readTree(
                        Files.newInputStream(
                                Paths.get("test_fixtures//SubFormFieldsPerDocument.json")));
        Iterator<Map.Entry<String, JsonNode>> fields = content.fields();

        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> kv = fields.next();
            String fieldName = kv.getKey();
            JsonNode expected = kv.getValue();

            SubFormFieldsPerDocumentBase base =
                    SubFormFieldsPerDocumentBase.init(expected.toString());

            assertTrue(Class.forName(packageNamePrefix + fieldName).isInstance(base));

            String serialized = mapper.writeValueAsString(base);
            JsonNode actual = mapper.readTree(serialized);

            // String comparison doesn't work due to json fields may be out of order
            assertEquals(expected, actual);
        }
    }

    @Test
    public void testSignersAllowsInt() throws Exception {
        ObjectMapper mapper = JSON.getDefault().getMapper();
        JsonNode content =
                mapper.readTree(
                        Files.newInputStream(
                                Paths.get("test_fixtures//SubFormFieldsPerDocument.json")));
        Iterator<Map.Entry<String, JsonNode>> fields = content.fields();

        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> kv = fields.next();
            JsonNode data = kv.getValue();
            String expected_signer = "1234";

            ((ObjectNode) data).put("signer", 1234);

            SubFormFieldsPerDocumentBase result =
                    SubFormFieldsPerDocumentBase.init(data.toString());

            assertEquals(expected_signer, result.getSigner());
        }
    }

    @Test
    public void testSignersAllowsString() throws Exception {
        ObjectMapper mapper = JSON.getDefault().getMapper();
        JsonNode content =
                mapper.readTree(
                        Files.newInputStream(
                                Paths.get("test_fixtures//SubFormFieldsPerDocument.json")));
        Iterator<Map.Entry<String, JsonNode>> fields = content.fields();

        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> kv = fields.next();
            JsonNode data = kv.getValue();
            String expected_signer = "sender";

            ((ObjectNode) data).put("signer", "sender");

            SubFormFieldsPerDocumentBase result =
                    SubFormFieldsPerDocumentBase.init(data.toString());

            assertEquals(expected_signer, result.getSigner());
        }
    }
}
